### ISSUE TITLE
unique is not working consistently

### ISSUE THREAD

**Issue by omriattiya** [Automattic/mongoose#7575]
I have Schema defined as:
````
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

var UsersAndRolesSchema = new Schema({
    roleName: {type:String, unique:true},
    userEmail: [{type:String}],
    children: [{type:Schema.Types.ObjectId, ref:'UsersAndRolesSchema'}]
});

module.exports = mongoose.model('UsersAndRoles', UsersAndRolesSchema);
````

and my tests are:

```
let globalBefore = async function () {
    mongoose.set('useCreateIndex', true);
    await mongoose.connect('mongodb://localhost:27017/Tests', {useNewUrlParser: true});
};

let globalAfter = function () {
    mongoose.connection.close();
};

let globalAfterEach = function () {
    mongoose.connection.db.dropDatabase();
};


describe('1. addUsersAndRole', function () {

    before(globalBefore);
    afterEach(globalAfterEach);
    after(globalAfter);


    it('1.1 adds 1 valid role', function (done) {
        usersAndRolesController.addUsersAndRole('role1', ['a@b.c'], (err, userAndRole) => {
            if (err) done(err);
            else {
                assert.deepEqual(userAndRole.userEmail, ['a@b.c']);
                assert.deepEqual(userAndRole.children, []);
                assert.deepEqual(userAndRole.roleName, 'role1');
                done();
            }
        });
    });

    it('1.2 adds the same role', function (done) {
        usersAndRolesController.addUsersAndRole('role1', ['a@b.c'], (err, userAndRole) => {
            if (err) done(err);
            else usersAndRolesController.addUsersAndRole('role1', ['a@b.c'], (err, userAndRole2) => {
                if (err) done();
                else {
                    console.log(userAndRole);
                    console.log(userAndRole2);
                    done(new Error("should have failed"));
                }
            });
        });
    });
});
```

when I run all of test 1 I get:
1.1 pass
1.2 
```
{ userEmail: [ 'a@b.c' ],
  children: [],
  _id: 5c7a417949ad9e2588e220d0,
  roleName: 'role1',
  __v: 0 }
{ userEmail: [ 'a@b.c' ],
  children: [],
  _id: 5c7a417949ad9e2588e220d1,
  roleName: 'role1',
  __v: 0 }
Error: should have failed
```

but when I run 1.2 alone its all good. 
I cant find the reason for that.
Any help?

**Comment by chidiwilliams** [comment]
When you drop and recreate your database, the indexes specified in your schema need to be explicitly recreated with `Model.ensureIndexes()`. Since you are dropping your DB after each test, run the function before each test.

**Comment by omriattiya** [comment]
> When you drop and recreate your database, the indexes specified in your schema need to be explicitly recreated with `Model.ensureIndexes()`. Since you are dropping your DB after each test, run the function before each test.

@chidiwilliams  Can you be more specific? I'm not sure how to use it.

**Comment by omriattiya** [comment]
@chidiwilliams worked. Thank you!
By the way, isn't `ensureIndexes` deprecated?

**Comment by vkarpov15** [comment]
@omriattiya use Model.createIndexes() instead, essentially equivalent: https://mongoosejs.com/docs/api.html#model_Model.createIndexes .

Also, remember that unique is not a validator: https://mongoosejs.com/docs/validation.html#the-unique-option-is-not-a-validator

**Comment by framerate** [comment]
I'm not sure but I may be seeing the same issue? I erased the season collection before deploying code with `unique: true` as instructed and the first deploy seemed ok. The 2nd test I ended up with duplicate rows!

```
// make sure each user only gets one entry per season
seasonSchema.index({ owner: 1, template: 1 }, { unique: true })
```

![image](https://user-images.githubusercontent.com/74291/60291388-a8411300-98cf-11e9-81ac-2b79d0985ccc.png)

The saving code, if relevant, is pretty simple:

```
playerSeason = await new Season({
        owner: data.player,
        template: currentSeason._id
      })
      await playerSeason.save()
```

Any advice is appreciated. I thought I was following the docs pretty accurately here.
